### PR TITLE
Add warning for synchronized on value classes

### PR DESCRIPTION
**Pull request by nox213**
fix https://github.com/scala/scala3/issues/17493
Thanks to @mbovel for his collaboration during the Spree.

### CODE REVIEW

File: tests/neg/i17493.scala
Comment (hamzaremmal) [inline tests/neg/i17493.scala:1]: You should probably move this file to the `tests/warn` directory and remove the `-Xfatal-warnings` from the compiler's options

Comment (nox213) [reply tests/neg/i17493.scala:1]: [68c30d1](https://github.com/scala/scala3/pull/20301/commits/68c30d1efb51f745ef6994ebac55189584c431df)
fixed

**Review by hamzaremmal — COMMENTED** [review]
Can you also squash everything in a single commit ? 

Let's wait for @sjrd's review before merging it

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:1160]: This won't do the right thing for an `import`ed `synchronized`.

You shouldn't look at the context owner, but at the `prefix` of `tree.tpe`.

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:1339]: This should probably reuse `checkSynchronizedCallOnValue`. Otherwise explicit selections on `AnyVal` won't do the right thing, AFAICT.

**Review by sjrd — CHANGES_REQUESTED** [review]

**Comment by mbovel** [comment]
> Can you also squash everything in a single commit ?

What is the benefit of squashing before merging? I would usually just use the "Squash and merge" feature from Github.

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala
Comment (mbovel) [reply compiler/src/dotty/tools/dotc/typer/RefChecks.scala:1339]: We tried the following during the Spree:

```scala
//> using options -explain
class A(val s: String) extends AnyVal
def Test =
  val a = A("hello")
  a.synchronized { println("hello, world") } // warn
```

And this is already throws an error:

```
 ~/scala-snippets-5 scalac foo.scala
-- [E008] Not Found Error: foo.scala:5:4 ---------------------------------------
5 |  a.synchronized { println("hello, world") } // warn
  |  ^^^^^^^^^^^^^^
  |value synchronized is not a member of A, but could be made available as an extension method.
  |
  |The following import might fix the problem:
  |
  |  import scala.reflect.Selectable.reflectiveSelectable
  |
1 error found
```

Is this the case you have in mind?

Anyway, I guess  it might be nice to use the same logic both for `Select` and `Apply`.

**Comment by mbovel** [comment]
Thanks to @sjrd's comments and re-reading discussions on #17449, I think what we did here is not correct.

The call to `synchronized` is resolved to a call to `Predef.synchronized` (as @som-snytt already noted on the issue's description):

```scala
sbt:scala3> scalac -Xprint:typer -Yprint-debug tests/warn/i17493.scala
...
[[syntax trees at end of                     typer]] // tests/warn/i17493.scala
package <root>.this.<empty> {
  final class A(s: scala.this.Predef.String) extends <root>.this.scala.AnyVal.
    <init>() {
    val s: scala.this.Predef.String
    def g: scala.this.Unit(inf) =
      scala.this.Predef.synchronized[scala.this.Unit^(inf)](
        {
          scala.this.Predef.println("hello, world")
        }
      )
  }
...
```

So we should not warn about a call on a value.

However, we have an existing machinery for warning about top-level calls to `synchronized`, so we should probably just adapt it for when the call is inside a value class definition. I will open a separate PR with a new fix suggestion in a minute.

**Comment by nox213** [comment]
I see. The issue wasn't that simple than I expected.

**Pull request closed without merging**
